File: C/hardware/hil_write_example/hil_write_example.c
//////////////////////////////////////////////////////////////////
//
// hil_write_example.c - C file
//
// This example writes one sample immediately to two analog output channels
// and four digital output channels.
//
// This example demonstrates the use of the following functions:
//    hil_open
//    hil_set_digital_directions
//    hil_write
//    hil_close
//
// Copyright (C) 2008 Quanser Inc.
//////////////////////////////////////////////////////////////////
    
#include "hil_write_example.h"

int main(int argc, char* argv[])
{
    static const char board_type[]       = "q2_usb";
    static const char board_identifier[] = "0";
    static char       message[512];

    qsigaction_t action;
    t_card board;
    t_int  result;

    /* Prevent Ctrl+C from stopping the application so hil_close gets called */
    action.sa_handler = SIG_IGN;
    action.sa_flags   = 0;
    qsigemptyset(&action.sa_mask);

    qsigaction(SIGINT, &action, NULL);

    result = hil_open(board_type, board_identifier, &board);
    if (result == 0)
    {
        const t_uint32 analog_channels[]  = { 0, 1 };
        const t_uint32 digital_channels[] = { 0, 1, 2, 3 };

        #define NUM_ANALOG_CHANNELS     ARRAY_LENGTH(analog_channels)
        #define NUM_DIGITAL_CHANNELS    ARRAY_LENGTH(digital_channels)

        t_double  voltages[NUM_ANALOG_CHANNELS];
        t_boolean values[NUM_DIGITAL_CHANNELS];
        t_uint32 channel;

        for (channel = 0; channel < NUM_ANALOG_CHANNELS; channel++)
            voltages[channel] = (channel - 1.5);
        for (channel = 0; channel < NUM_DIGITAL_CHANNELS; channel++)
            values[channel] = channel & 1;

        printf("This example writes constant values to the first four analog and\n");
        printf("digital output channels. The values written are:\n\n");
        for (channel = 0; channel < NUM_ANALOG_CHANNELS; channel++)
            printf("    DAC[%d] = %4.1f V\n", analog_channels[channel], voltages[channel]);
        printf("\n");
        for (channel = 0; channel < NUM_DIGITAL_CHANNELS; channel++)
            printf("    DIG[%d] = %d\n", digital_channels[channel], values[channel]);

        result = hil_set_digital_directions(board, NULL, 0, digital_channels, NUM_DIGITAL_CHANNELS);
        if (result == 0)
        {
            result = hil_write(board, analog_channels, NUM_ANALOG_CHANNELS, NULL, 0, digital_channels, NUM_DIGITAL_CHANNELS,
                NULL, 0, voltages, NULL, values, NULL);
            if (result < 0)
            {
                msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
                printf("Unable to write channels. %s Error %d.\n", message, -result);
            }
        }
        else
        {
            msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
            printf("Unable to set digital directions. %s Error %d.\n", message, -result);
        }
            
        hil_close(board);
    }
    else
    {
        msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
        printf("Unable to open board. %s Error %d.\n", message, -result);
    }

    printf("\nPress Enter to continue.\n");
    getchar();

    return 0;
}

File: python/hardware/hil_write_example/hil_write_example.py
#################################################################
#
# hil_write_example.py - Python file
#
# This example writes one sample immediately to two analog output channels
# and four digital output channels.
#
# This example demonstrates the use of the following functions:
#    HIL.open
#    HIL.set_digital_directions
#    HIL.write
#    HIL.close
#
# Copyright (C) 2023 Quanser Inc.
#################################################################

from quanser.hardware import HIL, HILError, Clock
import array as arr

board_type = "q2_usb"
board_identifier = "0"

try:
    card = HIL()
    card.open(board_type, board_identifier)

    try:
        analog_channels  = arr.array("i", [0, 1])
        digital_channels = arr.array("i", [0, 1, 2, 3])

        num_analog_channels  = len(analog_channels)
        num_digital_channels = len(digital_channels)

        voltages = arr.array("d", [0.0] * num_analog_channels)
        values   = arr.array("b", [0] * num_digital_channels)

        for channel in range(num_analog_channels):
            voltages[channel] = (channel - 1.5)
        for channel in range(num_digital_channels):
            values[channel] = channel & 1

        print("This example writes constant values to the first four analog and");
        print("digital output channels. The values written are:\n");
        for channel in range(num_analog_channels):
            print("    DAC[%d] = %4.1f V" % (analog_channels[channel], voltages[channel]))
        print()
        for channel in range(num_digital_channels):
            print("    DIG[%d] = %d" % (digital_channels[channel], values[channel]))
        print()

        card.set_digital_directions(None, 0, digital_channels, num_digital_channels)

        card.write(analog_channels, num_analog_channels, 
                   None, 0,
                   digital_channels, num_digital_channels,
                   None, 0,
                   voltages, 
                   None, 
                   values, 
                   None)

    except HILError as ex:
        print("Unable to write channels. %s" % ex.get_error_message())

    finally:
        card.close()

except HILError as ex:
    print("Unable to open board. %s" % ex.get_error_message())

input("Press Enter to continue.")
exit(0)
